esp-hal/src/i2c/master/mod.rs

```diff
@@ -1918,48 +1898,19 @@ impl Driver<'_> {
         Ok(())
     }
 
-    #[cfg(not(any(esp32, esp32s2)))]
-    /// Reads all bytes from the RX FIFO.
-    fn read_all_from_fifo_blocking(&self, buffer: &mut [u8]) -> Result<(), Error> {
-        // Read bytes from FIFO
-        // FIXME: Handle case where less data has been provided by the slave than
-        // requested? Or is this prevented from a protocol perspective?
-        for byte in buffer.iter_mut() {
-            loop {
-                self.check_errors()?;
-
-                let reg = self.regs().fifo_st().read();
-                if reg.rxfifo_raddr().bits() != reg.rxfifo_waddr().bits() {
-                    break;
-                }
-            }
-
-            *byte = read_fifo(self.regs());
-        }
-
-        Ok(())
-    }
-
-    #[cfg(any(esp32, esp32s2))]
-    /// Reads all bytes from the RX FIFO.
-    fn read_all_from_fifo_blocking(&self, buffer: &mut [u8]) -> Result<(), Error> {
-        // on ESP32/ESP32-S2 we currently don't support I2C transactions larger than the
-        // FIFO apparently it would be possible by using non-fifo mode
-        // see https://github.com/espressif/arduino-esp32/blob/7e9afe8c5ed7b5bf29624a5cd6e07d431c027b97/cores/esp32/esp32-hal-i2c.c#L615
-
+    /// Reads from RX FIFO into the given buffer.
+    fn read_all_from_fifo(&self, buffer: &mut [u8]) -> Result<(), Error> {
+        // we don't support single I2C reads larger than the FIFO
         if buffer.len() > I2C_FIFO_SIZE {
             return Err(Error::FifoExceeded);
         }
 
-        // wait for completion - then we can just read the data from FIFO
-        // once we change to non-fifo mode to support larger transfers that
-        // won't work anymore
-        self.wait_for_completion_blocking(false)?;
-
         // Read bytes from FIFO
-        // FIXME: Handle case where less data has been provided by the slave than
-        // requested? Or is this prevented from a protocol perspective?
         for byte in buffer.iter_mut() {
+            if self.regs().sr().read().rxfifo_cnt() == 0 {
```

Comment:
This should be a single check outside of the loop, I think. (in the ballpark of `rxfifo_cnt() < buffer.len()`) unless the hardware does something weird again